net_state: Add description to top level YAML/JSON API

cathay4t opened the pull request:
Introducing `description` as top level key of YAML/JSON API holding a
String. For example:

```yml
description: test-description: "Abc"
```

Currently, this property will be ignored when applying or querying, no
backend will persist this property. It is designed for user to store
description for the whole YAML only for now.

By using `testlib.apply.apply_with_description()`, we could store:
`test_description: <string>` into `.descriptions`. The
`pytest ----dump-ai-train-yaml` will store the YAML file into `.states`
folder for training AI.

The `--dump-ai-train-yaml` will only store YAML files with top level
description and the output will only contains the difference between
desired state and current state.

tyll commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -24,7 +25,7 @@ def eth1(eth1_up):
     ifstate = eth1_up[Interface.KEY][0]
     ifstate[Interface.IPV6][InterfaceIPv6.ENABLED] = True
-    libnmstate.apply(eth1_up)
+    apply_with_description("Enable IPv6 on interface eth1", eth1_up)
```
It is a great idea to save the description in the state. To me it seems useful for the generic schema as well, so that (for now) Nmstate will just ignore the description. Later, it can be great to be able to report this back as well, if we are keeping track of state changes like using a git commit history. Then the descriptions can be used in the git commit messages. What are your thoughts about this?

cathay4t replied:
The interface level description is stored in NetworkManager `user` setting, so it will also report back in `libnmstate.show()`. For YAML top level description cannot be stored in NM. But it would be good for AI to understand the YAML as transaction instead of per-interface changes and also good for future backend supporting commit like apply. Let me do a demo PR allowing top level descriptions.

tyll replied:
Maybe let's just use the top-level description instead. Re-using the existing field seems wrong, since I consider it as the description of the interface and not the description of the state.

cathay4t replied:
Changed to use top level descriptions. Please review.

tyll commented on rust/src/python/libnmstate/schema.py:
```diff
@@ -552,3 +552,6 @@ class Mptcp:
     FLAG_FULLMESH = "fullmesh"
+
+class Descriptions:
+    KEY = "descriptions"
```
```suggestion
    KEY = "description"
```

IMHO this should be singular. It is just a single string with the description.

tyll commented on tests/integration/testlib/apply.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def apply_with_description(description, desired_state, *args, **kwargs):
+    desired_state[Descriptions.KEY] = {"test-description": description}
```
IMHO this should be considered a description of the state in imperative mood. Since a test might use several states, "test-description" is not accurate for a single state.

cathay4t replied:
Changed to a string describing the whole YAML.

liangwen12year commented on tests/integration/ethernet_mtu_test.py:
```diff
@@ -136,12 +149,15 @@ def eth1_up_with_mtu_1900(eth1_up):
     eth1_desired_state = desired_state[Interface.KEY][0]
     eth1_desired_state[Interface.MTU] = 1900
 
-    libnmstate.apply(desired_state)
```
If it is possible to change all the `libnmstate.apply()` to `apply_with_description()` in all the integration tests, I can do it in a followup PR.

cathay4t replied:
Yes. That's the ultimate goal.

liangwen12year commented:
I ran `pytest --dump-states` on the tests `test_increase_iface_mtu` once, but a list of unrelated yaml files was generated, can we avoid this situation ?   


```
[root@fcvm .states]# ls
[root@fcvm nmstate]# ./pt.sh -k "test_increase_iface_mtu" 
[root@fcvm .states]# ls
test_increase_iface_mtu-0m57a94p.yml  test_increase_iface_mtu-5e7dwuq_.yml  test_increase_iface_mtu-c811muya.yml  test_increase_iface_mtu-ihee5ovg.yml  test_increase_iface_mtu-xdtxurcp.yml
test_increase_iface_mtu-194ky208.yml  test_increase_iface_mtu-7yj62039.yml  test_increase_iface_mtu-czddab9r.yml  test_increase_iface_mtu-r0nox29z.yml  test_increase_iface_mtu-zxaxd9z2.yml
test_increase_iface_mtu-24xwalt9.yml  test_increase_iface_mtu-aczsgd_r.yml  test_increase_iface_mtu-dj13psb9.yml  test_increase_iface_mtu-si_kh7ci.yml
[root@fcvm .states]# grep description -r .
./test_increase_iface_mtu-0m57a94p.yml:descriptions:
./test_increase_iface_mtu-0m57a94p.yml:  test-description: Increase MTU of interface eth1 to 1900
```

cathay4t commented:
Those are for k8s-nmstate go binding to test their schema there.
But yes, I can add new option to limit this dump to state with description only.

cathay4t commented:
Please use `pytest --dump-states-with-description-only`. So only YAML with description will be dumped.

liangwen12year commented on rust/src/lib/net_state.rs:
```diff
@@ -1,8 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
 
-#[cfg(not(feature = "gen_conf"))]
-use std::collections::HashMap;
```
This should not be removed ? 

`rust_lint` nightly CI is failed with the error:
```
error[E0412]: cannot find type `HashMap` in this scope
   --> src/lib/net_state.rs:267:17
    |
267 |     ) -> Result<HashMap<String, Vec<(String, String)>>, NmstateError> {
    |                 ^^^^^^^ not found in this scope
    |
help: consider importing this struct
    |
3   + use std::collections::HashMap;
```

ffmancera replied:
+1

cathay4t commented:
Currently, some test case are using full network state instead of the change only parts, this would be hard for AI to understand we just want to change an MTU among other unrelated current state.

Let's put this on hold till I add python binding for `gen_diff()` API.

cathay4t commented:
Waiting PR https://github.com/nmstate/nmstate/pull/2688 to provide formatted YAML file.

The command line argument changed to `--dump-ai-train-yaml`. 

This is one of the output YAML files:

```yml
description: Change MTU of eth1 to 1400
interfaces:
- name: eth1
  type: ethernet
  state: up
  mtu: 1400
```

ffmancera reviewed (commented):
Please, address changes from Wen. There are some issues related to an import being dropped.

liangwen12year commented:
In the commit message , the example of the description should be updated since the `description` is a String. 

```yml
description: test-description: "Abc"
```

cathay4t commented:
> In the commit message , the example of the description should be updated since the `description` is a String.
> 
> ```yaml
> description: test-description: "Abc"
> ```

Fixed.

cathay4t committed:
net_state: Add description to top level YAML/JSON API

Introducing `description` as top level key of YAML/JSON API holding a
String. For example:

```yml
description: Abc
```

Currently, this property will be ignored when applying or querying, no
backend will persist this property. It is designed for user to store
description for the whole YAML only for now.

By using `testlib.apply.apply_with_description()`, we could store:
`test_description: <string>` into `.descriptions`. The
`pytest ----dump-ai-train-yaml` will store the YAML file into `.states`
folder for training AI.

The `--dump-ai-train-yaml` will only store YAML files with top level
description and the output will only contains the difference between
desired state and current state.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
> Please, address changes from Wen. There are some issues related to an import being dropped.

Fixed.

liangwen12year reviewed (approved):
@cathay4t Thanks for the patch. I tested it and I did not find problems.
